vllm/model_executor/models/qwen2_audio.py

```diff
@@ -0,0 +1,453 @@
+):
+    """Gets a processor for the given model name via HuggingFace.
+
+    Derived from `vllm.transformers_utils.image_processor.get_image_processor`.
+    """
+    # don't put this import at the top level
+    # it will call torch.cuda.device_count()
+    from transformers import AutoProcessor
+
+    try:
+        processor = AutoProcessor.from_pretrained(
+            processor_name,
+            *args,
+            trust_remote_code=trust_remote_code,
+            **kwargs)
+    except ValueError as e:
+        # If the error pertains to the processor class not existing or not
+        # currently being imported, suggest using the --trust-remote-code flag.
+        # Unlike AutoTokenizer, AutoProcessor does not separate such errors
+        if not trust_remote_code:
+            err_msg = (
+                "Failed to load the processor. If the processor is "
+                "a custom processor not yet available in the HuggingFace "
+                "transformers library, consider setting "
+                "`trust_remote_code=True` in LLM or using the "
+                "`--trust-remote-code` flag in the CLI.")
+            raise RuntimeError(err_msg) from e
+        else:
+            raise e
+
+    return processor
+
+
+cached_get_processor = lru_cache(get_processor)
+
+
+def _get_feat_extract_output_lengths(input_lengths: torch.LongTensor):
+    """
+    Computes the output length of the convolutional layers
+    and the output length of the audio encoder
+    """
+    input_lengths = (input_lengths - 1) // 2 + 1
+    output_lengths = (input_lengths - 2) // 2 + 1
+    return input_lengths, output_lengths
+
+
+def get_max_qwen2_audio_audio_tokens(ctx: InputContext) -> int:
+    max_source_position = (
+        ctx.model_config.hf_config.audio_config.max_source_positions)
+    output_lengths = (max_source_position - 2) // 2 + 1
+    return output_lengths
+
+
+def input_processor_for_qwen2_audio(
+        ctx: InputContext, inputs: DecoderOnlyInputs) -> DecoderOnlyInputs:
+    multi_modal_data = inputs.get("multi_modal_data")
+    if multi_modal_data is None or "audio" not in multi_modal_data:
+        return inputs
+    if len(multi_modal_data["audio"]) == 0:
+        return inputs
+    assert (isinstance(multi_modal_data['audio'], list)
+            and isinstance(multi_modal_data['audio'][0], tuple))
+
+    processor = cached_get_processor(ctx.model_config.model)
+    audios = [_[0] for _ in multi_modal_data['audio']]
+
+    audio_inputs = processor.feature_extractor(audios,
+                                               sampling_rate=16000,
+                                               return_attention_mask=True,
+                                               padding="max_length")
+
+    feature_attention_mask = audio_inputs.pop("attention_mask")
+    input_features = audio_inputs['input_features']
+
+    audio_feat_lengths, audio_output_lengths = _get_feat_extract_output_lengths(
+        feature_attention_mask.sum(-1))
```

Comment:
If possible, avoid calling the `feature_extractor` again (since you've already done it in `input_mapper`). You can refer to the input processor of Ultravox for an example of how to make this more efficient.

Reply:
Hi, @DarkLight1337 I already removed `feature_extractor` here.